File: Graph/JohnsonAlgorithm/JohnsonAlgorithm.cpp
#include <iostream>
#include <vector>
#include <limits>
#include <queue>
#include <functional>

const int INF = std::numeric_limits<int>::max();

std::vector<int> bellmanFord(const std::vector<std::vector<std::pair<int, int>>>& graph, int source) {
    int V = graph.size();
    std::vector<int> distance(V, INF);
    distance[source] = 0;

    for (int _ = 0; _ < V - 1; ++_) {
        for (int u = 0; u < V; ++u) {
            for (const auto& neighbor : graph[u]) {
                int v = neighbor.first;
                int weight = neighbor.second;
                if (distance[u] != INF && distance[u] + weight < distance[v]) {
                    distance[v] = distance[u] + weight;
                }
            }
        }
    }

    return distance;
}

std::vector<std::vector<int>> johnson(const std::vector<std::vector<std::pair<int, int>>>& graph) {
    int V = graph.size();

    // Модификация графа
    std::vector<std::vector<std::pair<int, int>>> modifiedGraph = {{{0, 0}}};
    modifiedGraph.insert(modifiedGraph.end(), graph.begin(), graph.end());

    std::vector<int> h = bellmanFord(modifiedGraph, 0);

    // Проверка наличия отрицательного цикла
    if (std::find(h.begin(), h.end(), INF) != h.end()) {
        std::cout << "Граф содержит отрицательный цикл\n";
        return {};
    }

    // Модификация весов рёбер
    for (int u = 0; u < V; ++u) {
        for (auto& neighbor : graph[u]) {
            int v = neighbor.first;
            int weight = neighbor.second;
            graph[u][v] = weight + h[u] - h[v];
        }
    }

    // Запуск алгоритма Дейкстры для каждой вершины
    std::vector<std::vector<int>> allShortestPaths;
    for (int u = 0; u < V; ++u) {
        std::vector<int> distance(V, INF);
        distance[u] = 0;
        std::priority_queue<std::pair<int, int>, std::vector<std::pair<int, int>>, std::greater<>> minHeap;
        minHeap.push({0, u});

        while (!minHeap.empty()) {
            int currentDistance = minHeap.top().first;
            int currentVertex = minHeap.top().second;
            minHeap.pop();

            if (currentDistance > distance[currentVertex]) {
                continue;
            }

            for (const auto& neighbor : graph[currentVertex]) {
                int v = neighbor.first;
                int weight = neighbor.second;
                if (distance[currentVertex] + weight < distance[v]) {
                    distance[v] = distance[currentVertex] + weight;
                    minHeap.push({distance[v], v});
                }
            }
        }

        // Восстановление исходных весов
        for (int v = 0; v < V; ++v) {
            if (distance[v] != INF) {
                distance[v] -= h[u] - h[v];
            }
        }

        allShortestPaths.push_back(distance);
    }

    return allShortestPaths;
}

int main() {
    std::vector<std::vector<std::pair<int, int>>> graph = {
        {{1, 3}, {3, 7}},
        {{2, 2}},
        {{0, 5}, {3, 1}},
        {}
    };

    std::vector<std::vector<int>> result = johnson(graph);

    if (!result.empty()) {
        std::cout << "Matrix of shortest paths:\n";
        for (const auto& row : result) {
            for (int value : row) {
                if (value == INF) {
                    std::cout << "INF ";
                } else {
                    std::cout << value << " ";
                }
            }
            std::cout << "\n";
        }
    }

    return 0;
}

File: Graph/JohnsonAlgorithm/JohnsonAlgorithm.py
import heapq

INF = float('inf')

def johnson(graph):
    V = len(graph)

    # Шаг 1: Добавляем вершину s и соединяем её со всеми оригинальными вершинами
    graph.append([(V, 0) for _ in range(V)])

    # Шаг 2: Применяем алгоритм Беллмана-Форда для нахождения кратчайших путей от вершины s
    dist_s = [INF] * (V + 1)
    dist_s[V] = 0

    for _ in range(V):
        for u in range(V + 1):
            for v, weight in graph[u]:
                if dist_s[u] != INF and dist_s[u] + weight < dist_s[v]:
                    dist_s[v] = dist_s[u] + weight

    # Проверка наличия отрицательных циклов
    for u in range(V):
        for v, weight in graph[u]:
            if dist_s[u] != INF and dist_s[u] + weight < dist_s[v]:
                raise ValueError("Граф содержит отрицательный цикл")

    # Шаг 3: Удаляем добавленную вершину s и корректируем веса рёбер графа
    graph.pop()
    for u in range(V):
        for i in range(len(graph[u])):
            v, weight = graph[u][i]
            graph[u][i] = (v, weight + dist_s[u] - dist_s[v])

    # Шаг 4: Применяем алгоритм Дейкстры для каждой вершины графа
    all_shortest_paths = []
    for u in range(V):
        dist_u = dijkstra(graph, u)
        all_shortest_paths.append(dist_u)

    return all_shortest_paths

def dijkstra(graph, start):
    V = len(graph)
    dist = [INF] * V
    dist[start] = 0
    heap = [(0, start)]

    while heap:
        current_dist, u = heapq.heappop(heap)

        if current_dist > dist[u]:
            continue

        for v, weight in graph[u]:
            if dist[u] + weight < dist[v]:
                dist[v] = dist[u] + weight
                heapq.heappush(heap, (dist[v], v))

    return dist

# Пример использования
graph = [
    [(1, -2), (2, 4)],
    [(2, 3), (3, 2)],
    [(0, -1)],
    []
]

result = johnson(graph)
print("Матрица кратчайших путей:")
for row in result:
    print(row)